### PR TITLE
fix(input): fix URL validation being too strict

### LINKED ISSUE #13528: URL_REGEXP is not accurate
@andrewaustin @petebacondarwin

This commit broke the URL validation: https://github.com/angular/angular.js/commit/ffb6b2fb56d9ffcb051284965dd538629ea9687a

For example this URL (which uses a curly brackets macro syntax and should be a perfectly valid URL):
```
http://ads.xxxxx.com/xxxxxx/?pid=1234&pos=c&r={TIMESTAMP}
```

console

Also check this
http://stackoverflow.com/questions/5984116/regular-expression-for-w3c-compliant-urls

### PR DESCRIPTION
**Pull request by gkalpak**
Background:
Prior to ffb6b2f, there was a bug in `URL_REGEXP`, trying to match the hostname as `\S+` (meaning
any non-space character). This resulted in never actually validating the structure of the URL (e.g.
segments such as port, path, query, fragment).
Then ffb6b2f and subsequently e4bb838 fixed that bug, but revealed `URL_REGEXP`'s "strictness" wrt
certain parts of the URL.

Since browsers are too lenient when it comes to URL validation, this commit relaxes the "strictness"
of `URL_REGEXP`, focusing more on the general structure, than on the specific characters allowed in
each segment.

Note 1: `URL_REGEXP` still seems to be stricter than browsers in some cases.
Note 2: Browsers don't always agree on what is a valid URL and what isn't.

Fixes #13528

### CODE REVIEW

**Comment by gkalpak** [comment]
For anyone interested, I used [this pen][1] to visuallize how each browser's validation compares to ours for the URLs used in the test.

[1]: https://codepen.io/ExpertSystem/pen/ZQQbVN?editors=001

File: test/ng/directive/inputSpec.js [hunk 1]
```diff
@@ -2534,23 +2534,116 @@ describe('input', function() {
...
-        expect(URL_REGEXP.test('r2-d2.c3-p0://localhost/foo')).toBe(true);
-        expect(URL_REGEXP.test('abc:/foo')).toBe(true);
-        expect(URL_REGEXP.test('http://example.com/path;path')).toBe(true);
-        expect(URL_REGEXP.test('http://example.com/[]$\'()*,~)')).toBe(true);
-        expect(URL_REGEXP.test('http:')).toBe(false);
-        expect(URL_REGEXP.test('a@B.c')).toBe(false);
-        expect(URL_REGEXP.test('a_B.c')).toBe(false);
-        expect(URL_REGEXP.test('0scheme://example.com')).toBe(false);
-        expect(URL_REGEXP.test('http://example.com:9999/``')).toBe(false);
+    ddescribe('URL_REGEXP', function() {
```
Comment (petebacondarwin) [inline test/ng/directive/inputSpec.js]: `ddescribe` - tut tut!

File: test/helpers/privateMocks.js [hunk 2]
```diff
@@ -11,7 +11,7 @@ function baseThey(msg, vals, spec, itFn) {
   var valsIsArray = angular.isArray(vals);
 
   angular.forEach(vals, function(val, key) {
-    var m = msg.replace(/\$prop/g, angular.toJson(valsIsArray ? val : key));
+    var m = msg.split('$prop').join(angular.toJson(valsIsArray ? val : key));
```
Comment (petebacondarwin) [inline test/helpers/privateMocks.js]: what is the motivation for this change?

**Comment by petebacondarwin** [comment]
The regex looks good to me. Couple of minor comments to address.

File: test/ng/directive/inputSpec.js (hunk 1 above)
Comment (gkalpak) [inline test/ng/directive/inputSpec.js]: Was testing our CI :stuck_out_tongue:

File: test/helpers/privateMocks.js (hunk 2 above)
Comment (gkalpak) [inline test/helpers/privateMocks.js]: It's described in the commit message of the [first commit][1].

Basically, if your replacement (`val` or `key`) contains `$&`, it will be replaced with `$prop`.
Eg. `'Testing $prop'.replace(/\$prop/g, 'string with $&') === 'Testing string with $prop'`.

[1]: https://github.com/gkalpak/angular.js/commit/031303d3271905faf265470a6effe12ee7eebf6e

Comment (gkalpak) [inline test/helpers/privateMocks.js]: I just realized there a spec for the privateMocks. I'll add a test.

**Comment by gkalpak** [comment]
@petebacondarwin, I added a test to the first commit. PTAL

**Comment by petebacondarwin** [comment]
LGTM
